pkg/kubelet/server/stats/prometheus_resource_metrics.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package stats
+
+import (
+	"time"
+
+	"k8s.io/klog"
+	stats "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type NodeResourceMetric struct {
+	Name        string
+	Description string
+	ValueFn     func(stats.NodeStats) (*float64, time.Time)
+}
+
+func (n *NodeResourceMetric) desc() *prometheus.Desc {
+	return prometheus.NewDesc(
+		prometheus.BuildFQName("", "", n.Name),
+		n.Description, []string{}, nil)
+}
+
+type PodResourceMetric struct {
+	Name        string
+	Description string
+	ValueFn     func(stats.PodStats) (*float64, time.Time)
+}
+
+func (n *PodResourceMetric) desc() *prometheus.Desc {
+	return prometheus.NewDesc(
+		prometheus.BuildFQName("", "", n.Name),
+		n.Description, []string{"pod", "namespace"}, nil)
+}
+
+type ContainerResourceMetric struct {
+	Name        string
+	Description string
+	ValueFn     func(stats.ContainerStats) (*float64, time.Time)
+}
+
+func (n *ContainerResourceMetric) desc() *prometheus.Desc {
+	return prometheus.NewDesc(
+		prometheus.BuildFQName("", "", n.Name),
+		n.Description, []string{"container", "pod", "namespace"}, nil)
+}
+
+type ResourceMetricsConfig struct {
+	NodeMetrics      []NodeResourceMetric
+	PodMetrics       []PodResourceMetric
+	ContainerMetrics []ContainerResourceMetric
+}
+
+func NewPrometheusCoreCollector(provider SummaryProvider, config ResourceMetricsConfig) *resourceMetricCollector {
+	return &resourceMetricCollector{
+		provider: provider,
+		config:   config,
+		errors: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "scrape_error",
+			Help: "1 if there was an error while getting container metrics, 0 otherwise",
+		}),
+	}
+}
+
+type resourceMetricCollector struct {
+	provider SummaryProvider
+	config   ResourceMetricsConfig
+	errors   prometheus.Gauge
+}
+
+func (rc *resourceMetricCollector) Describe(ch chan<- *prometheus.Desc) {
+	rc.errors.Describe(ch)
+	for _, metric := range rc.config.NodeMetrics {
+		ch <- metric.desc()
+	}
+	for _, metric := range rc.config.PodMetrics {
+		ch <- metric.desc()
+	}
+	for _, metric := range rc.config.ContainerMetrics {
+		ch <- metric.desc()
+	}
+}
+
+func (rc *resourceMetricCollector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
ah, there's a reason we don't use normal prometheus collectors, which is that with pods and containers frequently starting and stopping, it's easy to "leak" metric collectors for those, which leads to all the expected issues with a reference leak.  Instead, we use this pattern, since at an given point in time we can easily answer the question "what are all the metrics available".

Reply:
added

Reply:
added a comment